File: articles/calculus-for-systems-modeling/articles/phase-lines-phase-planes-and-phase-portraits/fortran/phase_portrait_audit.f90
program phase_portrait_audit
  implicit none
  real(8) :: alpha, beta, delta, gamma, x, y, dxdt, dydt, speed
  integer :: xi, yi
  alpha = 0.7d0; beta = 0.05d0; delta = 0.02d0; gamma = 0.5d0
  print '(A)', 'x y dxdt dydt x_nullcline_residual y_nullcline_residual speed'
  do xi = 0, 60, 5
    do yi = 0, 30, 3
      x = dble(xi); y = dble(yi)
      dxdt = alpha*x - beta*x*y
      dydt = delta*x*y - gamma*y
      speed = sqrt(dxdt*dxdt + dydt*dydt)
      print '(7F12.6)', x, y, dxdt, dydt, dxdt, dydt, speed
    end do
  end do
end program

File: articles/calculus-for-systems-modeling/articles/phase-lines-phase-planes-and-phase-portraits/cpp/phase_portrait_audit.cpp
#include <cmath>
#include <iomanip>
#include <iostream>
#include <utility>

std::pair<double,double> rates(double x, double y, double alpha, double beta, double delta, double gamma){
  return {alpha*x - beta*x*y, delta*x*y - gamma*y};
}

int main(){
  double alpha = 0.7, beta = 0.05, delta = 0.02, gamma = 0.5;
  std::cout << std::fixed << std::setprecision(6);
  std::cout << "x,y,dxdt,dydt,x_nullcline_residual,y_nullcline_residual,speed,warning\n";
  for(int xi=0; xi<=60; xi+=5){
    for(int yi=0; yi<=30; yi+=3){
      double x = static_cast<double>(xi);
      double y = static_cast<double>(yi);
      auto [dxdt, dydt] = rates(x, y, alpha, beta, delta, gamma);
      double speed = std::sqrt(dxdt*dxdt + dydt*dydt);
      std::cout << x << "," << y << "," << dxdt << "," << dydt << "," << dxdt << "," << dydt << "," << speed << ",Vector-field values depend on parameter values state ranges and the assumed interaction structure.\n";
    }
  }
}
